### PR TITLE
miner: support custom block collation strategies

### PR DESCRIPTION
**Pull request by jwasinger**
depends on #23256.

This PR changes Geth's miner to allow for implementation of custom block collation strategies as Go plugins. A block collator receives empty pending blocks from the miner which it can fill with transactions from the pool (or another source), and commit to be the current sealing/pending block.  Plugins can also register custom RPC methods.

### Usage

`./geth --miner.collator /path/to/plugin --miner.collatorconfig /path/to/custom/config.toml`

See https://github.com/jwasinger/geth-mev-plugin for an example which re-implements https://github.com/flashbots/mev-geth miner changes as a plugin.

### Collator Plugins

Collator plugins are [Go plugins](https://pkg.go.dev/plugin).  Geth expects a plugin to export a plugin constructor method:

```
func PluginConstructor(config *map[string]interface{}, stack *node.Node) (Collator, error)
```

`config` represents an optional custom plugin toml configuration file.  A `Collator` implements custom block-filling strategies:

```
type Collator interface {
    CollateBlocks(miner MinerState, pool Pool, blockCh <-chan BlockCollatorWork, exitCh <-chan struct{})
    CollateBlock(bs BlockState, pool Pool)
}
```

`CollateBlocks` is a function that should run for the lifetime of the client (until it sees that `exitCh` is closed).  It is invoked in its own go-routine.  Every time a new empty pending block is created, it will appear for the collator to read from `blockCh`.  The `BlockCollatorWork` object read from `blockCh` has an empty `BlockState` for constructing and committing new blocks for sealing as well as a `Context` that the client will use to communicate when that block is stale due to the arrival of a new canonical chain head block.  A block is constructed by adding transactions from the pool (or another custom source) to a `BlockState` object and calling `Commit` (which sets the block as a the currently-sealing block).

`CollateBlock` is the hook for post-merge proposal of a single block.  The final call to `Commit` on the passed `BlockState` instance (or one of its copies) during the invocation of `CollateBlock` chooses the block that will be proposed.  Post-merge, `CollateBlocks` is used as a hook for receiving empty pending blocks. Post-merge, calling `Commit` on the `BlockState` instances passed to it is a no-op.

The `BlockState` API allows for adding transactions to a block, inspecting the state between transactions, deep-copying, and a `Commit` method.

#### Custom Plugin RPC Methods

The plugin constructor can choose to register custom rpc methods using the given `node.Node` object's `RegisterAPIs` method.

### CODE REVIEW

**Review by Sintay1994 — APPROVED** [review]

File: miner/collator.go [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+	// this makes sure multiple copies of a blockState can only trigger
+	// adjustment of the recommit interval once
+	interruptHandled *int32
+	// prevents calls to worker.commit (with a given blockState) after
+	// CollateBlock call on that blockState returns. examined in commit
+	// when commitMu is held.  modified right after CollateBlock returns
+	done *bool
+	// calling Commit() copies the value of env to this value
+	// and forwards it to the sealer via worker.commit() if shouldSeal is true
+	resultEnv *environment
```
Comment (rjl493456442) [inline miner/collator.go:217]: Is this resultEnv necessary? 
Whenever we create a blockState, we pass the generated environment to the collator and do whatever change it wants. Then we always send the deep copied environment to the commit function seems it will do some mutations inside. Finally we can either keep the env or drop it, but we don't need another field `resultEnv`.

File: miner/collator.go [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+	}
+
+	// Check whether the tx is replay protected. If we're not in the EIP155 hf
+	// phase, start ignoring the sender until we do.
+	if tx.Protected() && !bs.worker.chainConfig.IsEIP155(bs.env.header.Number) {
+		return ErrTxTypeNotSupported, nil
+	}
+
+	// TODO can this error also be returned by commitTransaction below?
+	_, err := tx.EffectiveGasTip(bs.env.header.BaseFee)
```
Comment (rjl493456442) [inline miner/collator.go:272]: It will be checked by `commitTransaction`, isn't it? https://github.com/ethereum/go-ethereum/blob/master/core/state_transition.go#L250

File: miner/collator.go (hunk 1 above)
Comment (rjl493456442) [reply miner/collator.go:217]: Perhaps in the collator, they will copy the block state and do some executions on the top and then pick a suitable one as the return? If so it makes sense to have the `resultEnv` field.

File: miner/collator.go [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+	logs       []*types.Log
+	shouldSeal bool
+	snapshots  []int
+	committed  bool
+
+	// shared values between multiple copies of a blockState
+
+	interrupt *int32
+	// mutex to make sure only one blockState is calling commit at a given time
+	commitMu *sync.Mutex
```
[the remark this answers is not in the document]
Comment (jwasinger) [reply miner/collator.go:207]: It's a pointer because `commitMu` can be shared among multiple copies of a blockState instance from a given work cycle.

**Comment by jwasinger** [comment]
The goal with these [shared contextual values](https://github.com/jwasinger/go-ethereum/blob/collator-v3/miner/collator.go#L203-L217) for the given work cycle was to give the following properties:

* calling Commit on a `BlockState` instance or its copies only interrupts the current sealing block if the call to CollateBlock where the instance was originally passed to has not returned.
* if a new canon chain head is received and triggers the miner interrupt while CollateBlock is executing, subsequent calls to Commit() or AddTransactions() for the passed `BlockState` instance (or any copies of it) for the given work cycle will fail and return an error.
 * if the recommit interval elapses and triggers the miner interrupt while `CollateBlock` is executing, subsequent calls to AddTransactions() for the passed `BlockState` instance (or copies of it) will fail  and return an error.
 * The recommit interval adjustment should only be triggered once per work cycle.
 
 But as I think about it, the implementation seems really hacky:

* Each call to `Commit` overwrites `resultEnv` , and its value becomes the pending block after `CollateBlock` returns when sealing.  But it seems unintuitive for it to be part of the `blockState`.
* The "shared contextual values" mentioned above really belong to the environment for the current work-cycle and not to `blockState`.
* Some block-level mutable values (`txs`) are in `environment` whereas others are in `blockState` which is not intuitive.  Anything that is unique to a given block should be in `blockState`.  The pending block should belong to the `environment` for that cycle.

These are just my high-level thoughts about making the changes more intuitive.  Going to see if I encounter issues trying to do a refactor.

File: miner/collator.go (hunk 1 above)
Comment (jwasinger) [reply miner/collator.go:217]: My thinking with `resultEnv` is that as a collator `Commit()`s various blocks, the `blockState`s track the last block committed to, and after `CollateBlock`, sets the pending block as `resultEnv`.

The pending block could be set directly as part of `blockState.Commit()` but it could cause the pending block to change erratically during `CollateBlock`.  That being said, I recall hearing that the pending block isn't really used that much so perhaps it's not worth caring about this behavior.

File: cmd/utils/flags.go [hunk 4]
```diff
@@ -477,6 +477,19 @@ var (
 		Name:  "miner.noverify",
 		Usage: "Disable remote sealing verification",
 	}
+	MinerCollatorPluginEnableFlag = cli.BoolFlag{
+		Name:  "miner.enablecollatorplugin",
```
[the remark this answers is not in the document]
Comment (karalabe) [reply cmd/utils/flags.go:481]: I don't think this flag is needed. We can use `--miner.collator=/path/to/lib`. It a collator is specified, enable it, if not, don't.

File: cmd/utils/flags.go [hunk 5]
```diff
@@ -477,6 +477,19 @@ var (
...
+		Usage: "Enable custom miner collator",
+	}
+	MinerCollatorPluginPath = cli.StringFlag{
+		Name:  "miner.collatorpluginfile",
```
[the remark this answers is not in the document]
Comment (karalabe) [reply cmd/utils/flags.go:485]: `--miner.collator` is fine imho

File: miner/collator.go [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+	InterruptNewHead
+)
+
+/*
+InterruptContext allows for active polling to detect if a new canonical chain
+head was received or recommit timer elapse occured.
+*/
+type InterruptContext interface {
+	InterruptState() int
+}
```
Comment (karalabe) [inline miner/collator.go:109]: I don't think this is needed. Using a `context.Context` is a more natural thing in Go world. I do see that you have 2 different interrupt types, but I don't think it's important from the collator's perspective. It should either stop generating, or should not interrupt imho.

I guess you're trying to somehow have the outside miner direct retrying the same block with a new set of txs. That perhaps should be left to the collator? I.e. if the collator wants, it can keep running and remaking blocks every 3/10/whatnot seconds based on the potentially new contents of the txs. That would keep this feature collator specific (e.g. it doesn't make sense t resubmit with a mev code), and we could get rid of the API complexity.

File: miner/collator.go [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+	}
+
+	bs.commitMu.Lock()
+	defer bs.commitMu.Unlock()
+	if *bs.done {
+		return false
+	}
+	if bs.shouldSeal {
+		bs.worker.commit(bs.env.copy(), bs.worker.fullTaskHook, true, bs.start)
+	}
```
Comment (karalabe) [inline miner/collator.go:364]: This checking for `done` and based on it calling or not calling `worker.commit` seems out of scope. If you already hav the result here, it's simpler to just always call commit and have that potentially reject it. It stops a lot of API leak across boundaries.

File: miner/collator.go [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+	Locals() []common.Address
+}
+
+type ReadOnlyState interface {
+	GetBalance(addr common.Address) *big.Int
+}
+
+type BlockState interface {
+	AddTransaction(tx *types.Transaction) (error, *types.Receipt)
+	RevertTransaction() bool
```
[the remark this answers is not in the document]
Comment (jwasinger) [reply miner/collator.go]: Removed `RevertTransaction` because we cannot revert multiple transactions with `StateDB`.

File: cmd/utils/flags.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jwasinger) [reply cmd/utils/flags.go]: done.

**Comment by holiman** [comment]
This was a good project, but ultimately did not gain the traction we had hoped -- having it be the base used for MEV so that MEV would not have to fork geth, but instead only build plugins.
We're in a different situation now, and I think this PR is no longer relevant. Time to close?

**Comment by jwasinger** [comment]
Yep
